packages/concerto-core/lib/serializer/jsonpopulator.js

```diff
@@ -77,6 +77,51 @@ function validateProperties(properties, classDeclaration) {
     }
 }
 
+/**
+ * Resolves the fully-qualified model name for a JSON object.
+ * @param {string} clazz the type name (FQN or short)
+ * @param {*} field a Field (which could be a Relationship)
+ * @returns {string} the fully qualified name of the object
+ * @throws {Error} if a short type name has not been imported
+ */
+function qualifyTypeName(clazz, field) {
+    const ns = ModelUtil.getNamespace(clazz);
+    if(ns.length > 0) {
+        return clazz; // already FQN
+    }
+    else {
+        // a short name, we use the namespace from the type of the field
+        const fqn = field.getFullyQualifiedTypeName();
+        return ModelUtil.getFullyQualifiedName(ModelUtil.getNamespace(fqn), clazz);
+    }
+}
+
+/**
+ * Resolves the fully-qualified model name for a JSON object.
+ * @param {*} obj an object with an optional $class
+ * @param {*} field a Field (which could be a Relationship)
+ * @returns {string} the fully qualified name of the object, based on its explicit $class
+ * or a $class inferred from the model
+ */
+function resolveFullyQualifiedTypeName(obj, field) {
+    if(obj.$class) {
+        return qualifyTypeName(obj.$class, field);
+    }
+    else {
+        const fqn = field.getFullyQualifiedTypeName();
+        const mm = field.getParent().getModelFile().getModelManager();
+        const classDecl = mm.getType(fqn);
+        const assignable = classDecl.isClassDeclaration?.() ? classDecl.getAssignableClassDeclarations()
+            .filter( a => !a.isAbstract()) : [classDecl];
+        if(assignable.length !== 1) {
+            throw new Error(`The type ${fqn} which was unambigious is now ambigious due to ${assignable.map(a => a.getFullyQualifiedName()).join(',')}`);
+        }
```

Comment:
This case worries me. I think we shouldn't omit $class in any case where it could become ambiguous in the future.